Added kubectl create role command

xingzhou opened the pull request:
Added `kubectl create role` command.

Fixed part of #39596 

**Release note**:
```
   Added one new command `kubectl create role` to help user create a single role from command line.
```

xingzhou committed:
Added kubectl create role command.

Added kubectl create role command.

xingzhou committed:
Added unit tests for kubectl create role command.

Added unit tests for kubectl create role command.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39852)
<!-- Reviewable:end -->

deads2k commented:
@fabianofranz The UI interaction for this is going to be important.  We don't want it to be surprising.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,88 @@
+// Role is a command to ease creating Roles.
+func NewCmdCreateRole(f cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "role NAME --verb=verb --resource=resource [--api-group=apigroup] [--resource-name=resourcename] [--non-resource-url=nonresourceurl] [--dry-run]",
```
When you run `kubectl` commands, you can actually do `kubect get resource.group`.  We could wire this command up to do the same thing so that I can do something like this: `kubectl create role foo --verb=get,list,watch --resource=jobs.batch,deployments.extensions,rs.extensions` (note the use of the shortcut) and come out the other end with a working role.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,88 @@
+	cmd.Flags().StringSlice("verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSlice("api-group", []string{}, "name of the APIGroup that contains the resources")
+	cmd.Flags().StringSlice("resource-name", []string{}, "resource in the white list that the rule applies to")
```
validation should probably stop you from specifying a resource name with more than one resource.

xingzhou replied:
Did not catch you here, do yo mean that a resource name option can not specify more than one resources? e.g. `--resource-name=a,b,c`

deads2k replied:
> Did not catch you here, do yo mean that a resource name option can not specify more than one resources? e.g. --resource-name=a,b,c

If I try to make a role with resources=rc,svc names=foo it will allow rc.name=foo and svc.name=foo.  It's almost never what someone wants.

deads2k replied:
The API (server-side) needs to be that flexible, but the CLI tool doesn't need to be.  Solve the 90% case and let them use json or yaml for the last 10%

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,88 @@
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSlice("api-group", []string{}, "name of the APIGroup that contains the resources")
+	cmd.Flags().StringSlice("resource-name", []string{}, "resource in the white list that the rule applies to")
+	cmd.Flags().StringSlice("non-resource-url", []string{}, "partial url that a user should have access to")
```
non-resource-url is only reasonable for clusterroles.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,88 @@
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.RoleV1GeneratorName)
+	cmd.Flags().StringSlice("verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
```
Is there any validation on this?  Using discovery to determine which resources are known would really cut down on typo related problems.  That means the validation is live only.

xingzhou replied:
Seems the server side does not take any validation on the resources yet. But let me try to add validation first from kubectl side.

xingzhou replied:
Also, for the validations, can we use RESTMapper's `ResourceFor` to validate resources?

deads2k replied:
> Seems the server side does not take any validation on the resources yet. But let me try to add validation first from kubectl side.

The server won't ever validate because you need to be able to create rules and synthentic rules, but the CLI only has to solve the 90% case and make it easy for people.

xingzhou replied:
got it, let me update the code then.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,88 @@
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.RoleV1GeneratorName)
+	cmd.Flags().StringSlice("verb", []string{}, "verb that applies to the resources contained in the rule")
```
I think we should limit the set of verbs allowed by default to limit typos.  That means we probably need an `--i-really-mean-it` flag.

xingzhou replied:
The same as resources, server side does not do validation on verbs. Let me add validation from kubectl side first.

janetkuo commented:
@k8s-bot ok to test

xingzhou commented:
@deads2k, I tried to update the code according to your comments, welcome your comments, thanks!

deads2k commented:
set command here to coordinate: https://github.com/kubernetes/kubernetes/pull/40059

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,89 @@
+
+var (
+	roleLong = templates.LongDesc(`
+		Create a Role to hold a logic grouping of permissions.`)
```
"Create a role with single rule"?

xingzhou replied:
sure, let me update it

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,89 @@
+func NewCmdCreateRole(f cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "role NAME --verb=verb --resource=resource [--api-group=apigroup] [--resource-name=resourcename] [--dry-run]",
+		Short:   "Create a Role to hold a logic grouping of permissions",
```
match long?

xingzhou replied:
yeap, except for the period in the end, but let me use long here

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,89 @@
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.RoleV1GeneratorName)
+	cmd.Flags().StringSlice("verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSlice("api-group", []string{}, "name of the APIGroup that contains the resources")
```
I thought this flag was going away in favor of supporting the "normal" kubectl behavior of `resource.group` in the resource arguments

deads2k replied:
By adding that behavior, you also gain the ability to have a rule which spans multiple API groups without accidentally granting permissions liek `pods.extensions`.

xingzhou replied:
If I understand correct, we also have to validate the match of resources and api groups, like you said, if user specified pods.extensions from the client, we need to tell them that resource `pods` is not in the api group `extensions`.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,89 @@
+		generator = &kubectl.RoleGeneratorV1{
+			Name:          name,
+			Verbs:         cmdutil.GetFlagStringSlice(cmd, "verb"),
+			Resources:     cmdutil.GetFlagStringSlice(cmd, "resource"),
```
@kargakis you're the generator expert.  Seems like the generator shouldn't deal with resolving shortnames like `rs.extensions` to `replicasets.extensions`, so this needs to be called differently.

kargakis replied:
I agree with you but the way the generator commands are structured now seems to indicate that generators handle validation of the command. If we structure the command as all cli commands that use the Complete-Validate-Run pattern then we can have resource resolution happen in Complete and run the generator in Run.

xingzhou replied:
Any suggestions on how we change this flag?

deads2k commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,304 @@
+var _ StructuredGenerator = &RoleGeneratorV1{}
+
+// Valid verb list for validation.
+var validVerbs []string = []string{"*", "get", "post", "put", "delete", "list", "create", "update", "patch", "watch", "proxy", "redirect", "deletecollection"}
```
split into valid resource version and valid non-resource verbs.  Resource verbs should probably be *, get, list, watch, update, patch, delete, deletecollection

deads2k commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,304 @@
+	containsResourceAll := false
+	for _, r := range resources {
+		// ResourceAll respresents all kinds of resources.
+		if r == "*" {
```
Let's not allow this by default. It's very bad practice.

xingzhou commented:
Updated the code according to the comments, in addition, added the short form name to long form name transfer, so now support the following kind of transfer: `rs.extensions` -> `replicasets.extensions`

deads2k commented on pkg/kubectl/kubectl.go:
```diff
@@ -197,6 +197,20 @@ func ResourceShortFormFor(resource string) (string, bool) {
 }
 
+// ResourceLongFormFor looks up for a long form of resource names.
+func ResourceLongFormFor(resource string) (string, bool) {
```
I think this would be more correct as a check using the `RESTMapper` we use since it includes the shortcut expander and this list will be legacy after we complete https://github.com/kubernetes/kubernetes/pull/40174 and related.

Also this doesn't handle the group correctly.

deads2k commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,264 @@
+		}
+		if resource.Resource != resourceName {
+			// Transfer from short form name to long form name.
+			longResourceName, _ := ResourceLongFormFor(resourceName)
```
@fabianofranz @kubernetes/sig-cli-misc this is vision sort of statement for generators.  I think this calculation should move toward the `Run` stage of `complete`, `validate`, `run` and the generation can remain dumb.

kargakis replied:
Yes, generators should not deal with restmapping, especially since we treat them as api and intend to move them server-side.

fabianofranz replied:
:+1:

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,87 @@
+// Role is a command to ease creating Roles.
+func NewCmdCreateRole(f cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "role NAME --verb=verb --resource=resource.group [--resource-name=resourcename] [--dry-run]",
```
I like the way this works. I think it will be quite helpful along with the related set command.

fabianofranz commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,87 @@
+		Create a role with single rule.`)
+
+	roleExample = templates.Examples(`
+		# Create a Role named "pod-reader" that allows user to perform "get", "watch" and "list" on "pods"
```
pods doesn't need quotes here

fabianofranz commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,87 @@
+
+	roleExample = templates.Examples(`
+		# Create a Role named "pod-reader" that allows user to perform "get", "watch" and "list" on "pods"
+		kubectl create role pod-reader --verb=get --verb=list --verb=watch --resource=pods`)
```
Can you also add an example with `--resource-name=`?

fabianofranz commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,264 @@
+// parseAndValidateVerbs parses and validates verbs.
+func (s *RoleGeneratorV1) parseAndValidateVerbs() error {
+	// support specify multiple verbs together
+	// e.g. --verb=get,watch,list
```
needs a test for this format.

fabianofranz commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,264 @@
+		if !arrayContains(validResourceVerbs, v) && !arrayContains(validNonResourceVerbs, v) {
+			return fmt.Errorf("invalid verb: '%s'", v)
+		}
+		if v == "*" {
```
Will this ever be reached, given the `if` above?

fabianofranz replied:
Needs a test for "*".

fabianofranz commented on pkg/kubectl/role.go:
```diff
@@ -0,0 +1,264 @@
+// parseAndValidateResourceNames parses and validates resource names.
+func (s *RoleGeneratorV1) parseAndValidateResourceNames() error {
+	// support specify multiple resource names together
+	// e.g. --resource-name=foo,boo
```
This needs a test.

xingzhou commented:
Made the following changes in the latest patch:
• Moved parsing part code to Complete method in create_role.go
• Added RunCreateRole method in create_role.go
• Added some unit test cases according to the comments

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+
+	// support specify multiple verbs together
+	// e.g. --verb=get,watch,list
+	verbs := cmdutil.GetFlagStringSlice(cmd, "verb")
```
bind these flag to variables when you register them above

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	// e.g. --verb=get,watch,list
+	verbs := cmdutil.GetFlagStringSlice(cmd, "verb")
+	for _, v := range verbs {
+		c.Verbs = mergeArrays(c.Verbs, strings.Split(v, ","))
```
If you bind `c.Verbs` directly to your stringslice, you could then remove duplicates in your run method so that delegation cases behave predictably and you don't have to have "get the flag and mess with it" logic here.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	candidateResources := []string{}
+	resources := cmdutil.GetFlagStringSlice(cmd, "resource")
+	for _, r := range resources {
+		candidateResources = mergeArrays(candidateResources, strings.Split(r, ","))
```
the flag stringslice should with commas for you in the "normal" way, whatever that is no-a-days.  Don't do something different with commas.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	for _, r := range resources {
+		candidateResources = mergeArrays(candidateResources, strings.Split(r, ","))
+	}
+	for _, r := range candidateResources {
```
add a comment explaining why you're doing this. (I know why, but for future maintainers)

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	}
+	for _, r := range candidateResources {
+		// support resource.group pattern
+		index := strings.Index(r, ".")
```
split on the dot

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+			resourceName, apiGroup = r[0:index], r[index+1:]
+		}
+
+		if !arrayContains(c.APIGroups, apiGroup) {
```
not quite.  You need to track `schema.GroupResource` tuples.  Not individual groups or resources.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+type CreateRoleOptions struct {
+	Name          string
+	Verbs         []string
+	Resources     []string
```
`Resources []schema.GroupResource` and eliminate `APIGroups`.

xingzhou replied:
But when we call generators, we still need to unpack the `GroupResource` to two string arrays as resources and api groups, since in generators, it will directly use string arrays and besides, the normal `Generate` API in generator requires string arrays, correct?

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+
+	// support specify multiple resource names together
+	// e.g. --resource-name=foo,boo
+	resourceNames := cmdutil.GetFlagStringSlice(cmd, "resource-name")
```
bind directly to this flag.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	// Transfer resource name from short form to long form.
+	resources := []string{}
+	for _, r := range role.Rules[0].Resources {
+		resource, err := mapper.ResourceFor(schema.GroupVersionResource{Resource: r})
```
run this logic before you run the generator passing in the `GroupResource` from your options to build up the resource. You may end up having to create multiple rules to support the command since `jobs.batch` and `hpa.autoscaling` require separate rules to express.

deads2k replied:
@derekwaynecarr @fabianofranz @kargakis this sort of logic really doesn't lend it self to a generator, but it could be done by passing in a slice of strings (we might have expanded shortcuts) and then having the generator parse the string and then build separate rules.  

Do we really want this kind of pain for what would be a pretty simple operation to do right here, especially knowing that in two years, we've *never* reused a generator?

xingzhou replied:
Do you mean that for `jobs.batch` and `hap.autoscaling`, we need to hard code checking them and create two rules? Why these two need to be taken special care of?

deads2k replied:
> xingzhou 13 hours ago Contributor
> Do you mean that for jobs.batch and hap.autoscaling, we need to hard code checking them and create two rules? Why these two need to be taken special care of?

Sorry, those aren't special, its just an example.  Because of the rule expansions, "get on (jobs,hpa) in (batch, autoscaling)" expands to 
 1. get on jobs.batch
 2. get on jobs.autoscaling
 3. get on hpa.batch
 4. get on hpa.autoscaling

That's an unexpected result.  You need to split the rule by API groups.

xingzhou replied:
Thanks for the clarification. IIUC, if we create separate rule for each of the api group here, that might behave differently as `kubectl create -f`. Since `create -f` now allows user to specify multiple api groups in a single rule. Do we need to apply the same policy to `create -f`?

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,228 @@
+	}
+	role.Rules[0].Resources = resources
+
+	gvks, _, err := typer.ObjectKinds(role)
```
@fabianofranz is there already or do you have plans to add a "create this `runtime.Object`" helper?

xingzhou commented:
Updated the patch according to the comments:
• Used schema.GroupVersionResource in create_role.go
• Kept using APIGroups and Resources arrays in role.go
• Created separate rule for each of the API Groups in role.go
• Removed comma option parsing code
Please take a look, thanks!

deads2k commented:
Alright, no one is clamouring on this thread to have generators and it's making the code harder.  Can you build the role directly in your run command and we'll comment from there?  I think it will make what I'm talking about with the tuple explosion more obvious.  RS.extensions and jobs.batch should create two rules, one for each.

xingzhou commented:
ok, I think the cause here is we want to validate the resource which requires the mapper object. Let me try to remove the generator and move all the validation and role making to the command itself.

xingzhou commented:
Updated the code with the following changes for discussion:
• Removed generator and moved all the validation and role creation code to the run command.
• Modified corresponding test cases.
Please take a look, thanks

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+		resourceName, apiGroup := "", ""
+
+		if len(sections) == 1 {
+			resourceName = r
```
just do the append directly in here and avoid having the `resourceName, apiGroup` variables.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+			verbs = append(verbs, v)
+		}
+		if v == "*" {
+			containsVerbAll = true
```
do this test first in the loop and break

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+	mapper, typer := f.Object()
+	dryRun, outputFormat := cmdutil.GetDryRunFlag(cmd), cmdutil.GetFlagString(cmd, "output")
+
+	// Remove duplicate verbs.
```
move this block to `Complete`

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+		verbs = []string{"*"}
+	}
+
+	// Remove duplicate resources and transfer resource name from short form to long form.
```
add doc describing the map.  I believe it is keyed by group pointing to mapped resource.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+	groupResourceMapping := map[string][]string{}
+	for _, r := range c.Resources {
+		resource, err := mapper.ResourceFor(r)
+		if err != nil {
```
document that this prevents pointing to non-existent resources and allows resource shortnames to be used in the CLI.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+		}
+	}
+
+	// Remove duplicate resource names.
```
Move to `Complete`

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+		return err
+	}
+	gvk := gvks[0]
+	mapping, err := mapper.RESTMapping(schema.GroupKind{Group: gvk.Group, Kind: gvk.Kind}, gvk.Version)
```
You know exactly what you're going to create.  Just call `f.ClientSet` and select the one you want.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return err
+	}
+	if err := kubectl.UpdateApplyAnnotation(info, f.JSONEncoder()); err != nil {
```
This call does nothing because the annotation isn't present.

deads2k commented on pkg/kubectl/cmd/create_role.go:
```diff
@@ -0,0 +1,255 @@
+
+	// Create separate rule for each of the api group.
+	rules := []rbac.PolicyRule{}
+	for g, resources := range groupResourceMapping {
```
use `sets.StringKeySet(map).List()` to force a consistent ordering.

deads2k commented:
Add a test-cmd test please.

xingzhou commented:
updated the code according to @deads2k latest comments, PTAL thx. BTW, for docs, do I need to run `hack/genreate-docs.sh` to generate the docs? but I saw this comments in the existing doc:
`This file is autogenerated, but we've stopped checking such files into the
repository to reduce the need for rebases. Please run hack/generate-docs.sh to
populate this file.`

deads2k commented:
/lgtm

deads2k commented:
> updated the code according to @deads2k latest comments, PTAL thx. BTW, for docs, do I need to run hack/genreate-docs.sh to generate the docs? but I saw this comments in the existing doc:
> This file is autogenerated, but we've stopped checking such files into the repository to reduce the need for rebases. Please run hack/generate-docs.sh to populate this file.

Yeah, I would run the script to satisfy the verifier.

deads2k commented:
> /lgtm

nice work.  Can you add a 
```
\```release-note
\```
```
block in your PR description.

xingzhou commented:
@deads2k, I tried to run `hack/generate-docs.sh` locally on my Mac. But seems the behavior of the script has changed. Besides creating create-role related docs and updating the `.generated_docs` file, it also changed the content of all the existing docs file. I can just add the create-role docs and `.generated_docs` file to this patch, or do we need to file a new ticket on this script?

kargakis commented:
@xingzhou for the purposes of merging code in Kube I think you should run `hack/update-generated-docs.sh`

deads2k commented:
> @xingzhou for the purposes of merging code in Kube I think you should run hack/update-generated-docs.sh

Try the generate into a separate commit and push.  If the bot likes it, we'll merge it.

xingzhou committed:
Added generated docs for kubectl create role command.

Added generated docs for kubectl create role command.

xingzhou commented:
@deads2k and @kargakis, generated the doc by using `hack/update-generated-docs.sh`, now seems the bot is happy, PTAL, thanks!

deads2k commented:
I'll tag this as-is this time (it's not egregious), but next time please squash.  If you get to it before it merges, please squash and tag me to tag again.

/lgtm
/approve

deads2k commented:
@eparis @apelisse special case the generated files?

xingzhou commented:
thanks David for the information. I'm now working on the create clusterrole command, will use 1 single commit then.

deads2k commented:
> thanks David for the information. I'm now working on the create clusterrole command, will use 1 single commit then.

Thanks.  Tag me into that.  I'm looking forward to it.
